eth/transactor.go

```diff
@@ -227,16 +205,116 @@ func (t *Transactor) transact(
+	// 1. Determine gas price
+	// Only accept legacy flags or EIP-1559 flags, not both
+	hasLegacyFlags := txopts.forceGasGwei > 0 || txopts.minGasGwei > 0 || txopts.maxGasGwei > 0 || txopts.addGasGwei > 0
+	has1559Flags := txopts.maxFeePerGasGwei > 0 || txopts.maxPriorityFeePerGasGwei > 0
+	if hasLegacyFlags && has1559Flags {
+		return ErrConflictingGasFlags
+	}
+	// Check if chain supports EIP-1559
+	ctx, cancel := context.WithTimeout(context.Background(), ctxTimeout)
+	defer cancel()
+	head, err := client.HeaderByNumber(ctx, nil)
+	if err != nil {
+		return fmt.Errorf("failed to call HeaderByNumber: %w", err)
+	}
+	send1559Tx := false
+	if head.BaseFee != nil && !hasLegacyFlags {
+		send1559Tx = true
+		err = determine1559GasPrice(ctx, signer, txopts, client, head)
+		if err != nil {
+			return fmt.Errorf("failed to determine EIP-1559 gas price: %w", err)
+		}
+	} else {
+		// Legacy pricing
+		err = determineLegacyGasPrice(ctx, signer, txopts, client)
+		if err != nil {
+			return fmt.Errorf("failed to determine legacy gas price: %w", err)
+		}
+	}
+
+	// 2. Determine gas limit
+	if txopts.gasLimit > 0 {
+		// Use the specified limit if set
+		signer.GasLimit = txopts.gasLimit
+		return nil
+	} else if txopts.addGasEstimateRatio > 0.0 {
+		// 2.1. Estimate gas
+		signer.NoSend = true
+		dryTx, err := method(client, signer)
+		if err != nil {
+			return fmt.Errorf("tx dry-run err: %w", err)
+		}
+		signer.NoSend = false
+		var typesMsg types.Message
+		londonSigner := types.NewLondonSigner(t.chainId)
+		if send1559Tx {
+			typesMsg, err = dryTx.AsMessage(londonSigner, head.BaseFee)
+		} else {
+			typesMsg, err = dryTx.AsMessage(londonSigner, nil)
+		}
+		if err != nil {
+			return fmt.Errorf("failed to get typesMsg: %w", err)
+		}
+		callMsg := ethereum.CallMsg{
+			From:     typesMsg.From(),
+			To:       typesMsg.To(),
+			GasPrice: typesMsg.GasPrice(),
+			Value:    typesMsg.Value(),
+			Data:     typesMsg.Data(),
+		}
+		estimatedGas, err := client.EstimateGas(ctx, callMsg)
+		if err != nil {
+			return fmt.Errorf("failed to call EstimateGas: %w", err)
+		}
+		// 2.2. Multiply gas limit by the configured ratio
+		signer.GasLimit = uint64(float64(estimatedGas) * (1 + txopts.addGasEstimateRatio))
+	}
+	// If addGasEstimateRatio not specified, just defer to go-ethereum for gas limit estimation
+	return nil
+}
+
+// determine1559GasPrice sets the gas price on the signer based on the EIP-1559 fee model
+func determine1559GasPrice(
+	ctx context.Context, signer *bind.TransactOpts, txopts txOptions, client *ethclient.Client, head *types.Header) error {
+	if txopts.maxPriorityFeePerGasGwei == 0 {
+		gasTipCap, suggestGasTipCapErr := client.SuggestGasTipCap(ctx)
+		if suggestGasTipCapErr != nil {
+			return fmt.Errorf("failed to call SuggestGasTipCap: %w", suggestGasTipCapErr)
+		}
+		signer.GasTipCap = gasTipCap
+	} else {
+		signer.GasTipCap = new(big.Int).SetUint64(txopts.maxPriorityFeePerGasGwei * 1e9)
+	}
+	if txopts.maxFeePerGasGwei == 0 {
```

Comment:
we should only need to specify GasTipCap in txopts, and geth code calculates txopts.GasFeeCap and creates DynamicFeeTx if GasFeeCap isn’t nil. 

what do we get by set maxFeePerGasGwei cap? we kind of know it doesn't really work. if we don't have this option, then we don't need this part of code and just let geth client handles it

Reply:
This is like the old MaxGasPrice if we absolutely want to limit the gas price, maybe for non-urgent tasks, so in case of a sudden spike in base fee we just wait.

Reply:
if txopts.maxFeePerGasGwei == 0 
instead of do geth logic ourselves, can you just return (signer.GasFeeCap is nil) so geth does the work?

I'm ok we keep maxfee opt for future flexibility, but in practice we shouldn't use it

Reply:
Ah yes, can also defer to geth if maxPriorityFeePerGasGwei is 0. Fixed to only set if specified.